### PR TITLE
init KVCache fix for grpo trainer

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

originally kvcache was initialized using `config.global_batch_size_to_train_on` and model was initialized using `max_utils.get_batch_seq_len_for_mode`. These ways assumed that the whole mesh is going to be used for model initialization. 

In grpo_trainer with disaggregated setup, we initialize the training and inference model with a submesh. This PR uses mesh.size to calculate global batch size used for initializing the model. 

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Comment by khatwanimohit** [comment]
> Thanks Mohit. For my understanding, can you please share the behavior you saw here? Was it throwing an exception/not being partitioned properly?

It was throwing an error at [attention_op.py](https://github.com/AI-Hypercomputer/maxtext/blob/19b9f7f86e55b09662220eb0d89567ae14476dc0/src/MaxText/layers/attention_op.py#L530) complaining the batch size of query and key, value don't match. 
That error helped me find the reason because we were initializing key, values batch size based on jax.devices()

File: src/MaxText/maxtext_utils.py [hunk 1]
```diff
@@ -1060,7 +1060,7 @@ def get_prefill_kv_cache_annotations(model, config, rng, mesh, page_state: None
 
   def init_kv_cache(model, config):
     input_shape = (
-        config.micro_batch_size_to_train_on,
+        int(config.per_device_batch_size * mesh.size),
```
Comment (bvandermoon) [inline src/MaxText/maxtext_utils.py:1063]: @gobbleturk is this the right value to use here?

**Comment by khatwanimohit** [comment]
Thank you @gobbleturk for suggesting to change in https://github.com/AI-Hypercomputer/maxtext/blob/07205e24f770501214735cf80ff6358d6c831915/src/MaxText/pyconfig.py#L1166 instead which made me realize we already support initializing a model on a submesh using config.subslice_shape. Closing this PR
